### PR TITLE
[QT-19] Enable Enos replication scenario

### PR DESCRIPTION
**Pull request by jaymalasinha**
Add Enos replication scenario to test Vault enterprise artifacts

### CODE REVIEW

**Comment by jaymalasinha** [comment]
Can be set by passing vault_unseal_when_no_init = "false"

File: enos/enos-modules.hcl
Comment (ryancragun) [inline enos/enos-modules.hcl:90]: This comment applied to many of the new modules. Generally we try to include a verb that describes what the module goes. For this I'd suggest something like `vault_get_cluster_ips`

File: enos/modules/vault_performance_replication_primary/main.tf
Comment (ryancragun) [inline enos/modules/vault_performance_replication_primary/main.tf:39]: Should we be capitalizing all the env vars?

File: enos/modules/vault_cluster_ips/main.tf
Comment (ryancragun) [inline enos/modules/vault_cluster_ips/main.tf:76]: Should we capitalize all of the env vars?

File: enos/modules/vault_performance_replication_primary/scripts/configure-vault-pr-primary.sh
Comment (ryancragun) [inline enos/modules/vault_performance_replication_primary/scripts/configure-vault-pr-primary.sh:22]:
```suggestion
# so we do not enable here.
```

File: enos/modules/vault_unseal_nodes/main.tf
Comment (ryancragun) [inline enos/modules/vault_unseal_nodes/main.tf:30]: should we add type and description information to this variable?

File: enos/modules/vault_unseal_nodes/main.tf
Comment (ryancragun) [inline enos/modules/vault_unseal_nodes/main.tf:60]: I think we should updated VAULT-12309 with acceptance criteria that includes removing this workaround, or creating a task in our backlog that is blocked by VAULT-12309 that removes this.

Comment (ryancragun) [reply enos/modules/vault_unseal_nodes/main.tf:60]: The same feedback applies to the next workarounds and VAULT-12311.

File: enos/modules/vault_unseal_nodes/scripts/unseal-node.sh
Comment (ryancragun) [inline enos/modules/vault_unseal_nodes/scripts/unseal-node.sh:8]: nit: inconsistent indentation

File: enos/modules/vault_verify_performance_replication/main.tf
Comment (ryancragun) [inline enos/modules/vault_verify_performance_replication/main.tf:40]: Should we remove these unused variables?

File: enos/modules/vault_verify_performance_replication/main.tf
Comment (ryancragun) [inline enos/modules/vault_verify_performance_replication/main.tf:79]: These were introduced in Terraform 1.2.0. Do we need to update documentation/config or add a minimum Terraform version requirement to this module?

File: enos/modules/vault_verify_performance_replication/main.tf
Comment (ryancragun) [inline enos/modules/vault_verify_performance_replication/main.tf:104]: 🏆

File: enos/modules/vault_verify_performance_replication/scripts/verify-performance-replication.sh
Comment (ryancragun) [inline enos/modules/vault_verify_performance_replication/scripts/verify-performance-replication.sh:36]: I'm not sure this makes sense. We know the cluster state is idle so why are we querying for it again?

File: enos/enos-variables.hcl
Comment (ryancragun) [inline enos/enos-variables.hcl:36]: Is there a particular reason we change the default region?

File: enos/enos-modules.hcl
Comment (ryancragun) [inline enos/enos-modules.hcl:96]: Can we add a comment and/or modify the name to indicate that this is a workaround for known issues? I wouldn't want future scenario authors to use it without understanding that we intend to remove it.

File: enos/enos-scenario-autopilot.hcl
Comment (ryancragun) [inline enos/enos-scenario-autopilot.hcl:140]: Love it. Adding more validation to the autopilot scenario is wonderful.

File: enos/enos-scenario-replication.hcl
Comment (ryancragun) [inline enos/enos-scenario-replication.hcl:1]: Can we add a comment with more context than just replication? I think a `description` field in a `scenario` that we can use in enos to automatically tell us that additional information would be great. I created QT-456 for it just now. Until then, a bit more context up front as a comment might be nice.

File: enos/enos-scenario-replication.hcl
Comment (ryancragun) [inline enos/enos-scenario-replication.hcl:338]: Can we add a comment on which steps are workarounds for bugs here as well as in the called modules?

File: enos/enos-scenario-replication.hcl
Comment (ryancragun) [inline enos/enos-scenario-replication.hcl:499]: ‼️ 👶 💉 ⬇️

File: enos/modules/vault_verify_unsealed/templates/verify-vault-node-unsealed.sh
Comment (ryancragun) [inline enos/modules/vault_verify_unsealed/templates/verify-vault-node-unsealed.sh:16]: Interesting! I guess the status endpoint doesn't need a token. Makes sense as you can run it before init.

File: enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh
Comment (ryancragun) [inline enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh:45]:
```suggestion
# Create new user and attach reguser policy
```

File: enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh
Comment (ryancragun) [inline enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh:42]: I don't have a problem with adding a regular user, I'm just curious why we do it for testing the k/v secrets engine.

**Review by ryancragun — COMMENTED** [review]
What a triumph. I can only imagine the pain of finding those two issues while whipping up this scenario. Great, great job.
 
It doesn't look like the replication scenario has been added into the Github Actions matrix yet. Hopefully that stuff has stabilized and it shouldn't be too much work to include the replication scenarios into the various matrices. I'd be happy to help out there.

File: enos/enos-modules.hcl
Comment (jaymalasinha) [reply enos/enos-modules.hcl:60]: Good point, will update the module name to `shutdown_host` 👍 ! Originally it was supposed to remove_host but had to change to `shutdown` to gracefully handle abrupt removal.

File: enos/modules/vault_cluster_ips/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_cluster_ips/main.tf:45]: 💯

File: enos/modules/vault_unseal_nodes/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_unseal_nodes/main.tf:39]: `wait_until_sealed` sounds better 👍

File: enos/enos-modules.hcl
Comment (jaymalasinha) [reply enos/enos-modules.hcl:96]: The module itself is a required step in replication scenario, where the secondary nodes need to be unsealed if they have the seal type "Shamir".  The workaround for the issue is to call a custom script twice to unseal, instead of using the `vault_unseal_resource`.  So I added the comments with reported issue links within this module and not on here.

File: enos/enos-scenario-replication.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-replication.hcl:1]: done

File: enos/enos-scenario-replication.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-replication.hcl:338]: The step itself is not a workaround, the method used within it is.. so added the comment within the module

File: enos/enos-variables.hcl
Comment (jaymalasinha) [reply enos/enos-variables.hcl:36]: Good catch! I had changed it when trying to setup replication between different zones, but didn't change it back. Will update it now.

File: enos/modules/vault_unseal_nodes/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_unseal_nodes/main.tf:60]: Will make a note to add it with links once this PR is merged

File: enos/modules/vault_verify_performance_replication/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_verify_performance_replication/main.tf:79]: We already have it https://github.com/hashicorp/vault/blob/main/enos/enos-terraform.hcl#L19

File: enos/modules/vault_verify_performance_replication/scripts/verify-performance-replication.sh
Comment (jaymalasinha) [reply enos/modules/vault_verify_performance_replication/scripts/verify-performance-replication.sh:36]: Updated to display the `cluster_state`

File: enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh
Comment (jaymalasinha) [reply enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh:42]: That's a great question, I'll add a comment clarifying this.  So we add a regular user here because we also use this module to verify data in replication scenario.  When replication is enabled, we cannot authenticate to the secondary server using the root token, so I add a user here so we can authenticate and verify written data for all scenarios

File: enos/enos-scenario-replication.hcl
Comment (joshbrand) [inline enos/enos-scenario-replication.hcl:129]: do these (and other steps) still need `vault` in the name? there's some inconsistency

File: enos/enos-scenario-replication.hcl
Comment (joshbrand) [inline enos/enos-scenario-replication.hcl:156]: 🤔 we might want to make this exposed via a variable (thinking ahead for longer-running scenarios, this could fill up our tiny default sized disks)

File: enos/enos-scenario-replication.hcl
Comment (joshbrand) [inline enos/enos-scenario-replication.hcl:498]: is the `_1` just because its a single follower?

File: enos/enos-scenario-upgrade.hcl
Comment (joshbrand) [inline enos/enos-scenario-upgrade.hcl:252]: something like "get_updated_cluster_ips" might be better?

File: enos/enos-modules.hcl
Comment (jaymalasinha) [reply enos/enos-modules.hcl]: Done

File: enos/modules/vault_cluster_ips/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_cluster_ips/main.tf]: Removed

File: enos/modules/vault_unseal_nodes/main.tf
Comment (jaymalasinha) [reply enos/modules/vault_unseal_nodes/main.tf]: updated

File: enos/modules/vault_unseal_nodes/scripts/wait-till-sealed.sh
Comment (jaymalasinha) [reply enos/modules/vault_unseal_nodes/scripts/wait-till-sealed.sh]: fixed

File: enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh
Comment (jaymalasinha) [reply enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh]: updated

File: enos/enos-scenario-replication.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-replication.hcl:39]: removed

File: enos/enos-scenario-replication.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-replication.hcl:163]: for consul it is a consul cluster, but you are correct, backend makes more sense 👍

File: enos/enos-scenario-replication.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-replication.hcl:498]: It is `follower_1`.  The `vault_get_cluster_ips` module identifies the nodes as `leader`, `follower_1` and `follower_2` so it facilitates removing the identified node in these steps

File: enos/enos-scenario-upgrade.hcl
Comment (jaymalasinha) [reply enos/enos-scenario-upgrade.hcl:252]: updated to `get_vault_cluster_ips` and `get_updated_vault_cluster_ips`

File: enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh
Comment (jaymalasinha) [reply enos/modules/vault_verify_write_data/scripts/smoke-enable-secrets-kv.sh:42]: added comments

File: enos/modules/vault_unseal_nodes/main.tf
Comment (mpalmi) [reply enos/modules/vault_unseal_nodes/main.tf:60]: Just a note here: this behavior (though likely a bug) does not have a huge impact from the end-user's perspective so it is unlikely to get prioritized anytime soon.

File: enos/modules/vault_unseal_nodes/scripts/unseal-node.sh
Comment (mpalmi) [inline enos/modules/vault_unseal_nodes/scripts/unseal-node.sh:12]: Shouldn't we only have to retry once here? My understanding was the first unseal attempt fails, so shamir unseals take N+1 tries (where N is the number of seals)

If all of that is correct, is there a way to use the existing unseal module twice and expect an error on the first attempt, or do we need this script?

If not, maybe we can simplify this script to do ^.

**Review by mpalmi — APPROVED** [review]
Looks great! Thanks for doing this @jaymalasinha! 👍 from me, but I left a few nits if you feel like addressing them (not blockers).

File: enos/modules/vault_unseal_nodes/scripts/unseal-node.sh
Comment (jaymalasinha) [reply enos/modules/vault_unseal_nodes/scripts/unseal-node.sh:12]: I am sure it exits in the second attempt to unseal, `5` here is just an upper limit after which it would fail with error unsealing

**Comment by jaymalasinha** [comment]
> Looks great! Thanks for doing this @jaymalasinha! 👍 from me, but I left a few nits if you feel like addressing them (not blockers).

Thank you @mpalmi for your help with debugging this scenario ❤️

**Review by joshbrand — APPROVED** [review]
LGTM!

**Pull request merged**
